test/pylib/cpp/boost/prepare_instance.py

```diff
@@ -0,0 +1,116 @@
+    """
+    def __init__(self, env_file: Path):
+        self.env_file = env_file
+
+    def prepare(self) -> None:
+        """
+        Read the environment variables for S3 and MinIO from the file and set them for the process.
+        """
+        timeout = 10
+        start_time = time.time()
+        while True:
+            if os.path.exists(self.env_file):
+                with open(self.env_file, 'r') as file:
+                    for line in file.readlines():
+                        key, value = line.strip().split('=', 1)
+                        os.environ[key] = value
+                break
+
+            if time.time() - start_time > timeout:
+                raise TimeoutError(f"Timeout waiting for file {self.env_file}")
+            # Sleep needed to wait when the controller will create a file with environment variables.
+            # Without sleep checking of the file existence will be too fast,
+            # so it will finish before the file is created
+            time.sleep(0.5)
+
+    def cleanup(self) -> None:
+        """
+        Fake method to have the same interfaces with Controller class.
+        """
+        pass
+
+
+class PrepareMainProcessEnv:
+    """
+    A class responsible for starting additional services needed by tests.
+
+    It starts up a Minio server and an S3 mock server.
+    The environment settings are saved to a file for later consumption by child processes.
+    """
+    def __init__(self, temp_dir: Path, modes: list[str], env_file: Path):
+        self.temp_dir = temp_dir
+        pytest_dirs = [self.temp_dir / mode / 'pytest' for mode in modes]
+        for directory in [self.temp_dir, *pytest_dirs]:
+            if not directory.exists():
+                directory.mkdir()
+        self.env_file = env_file
+        hosts = HostRegistry()
+        self.loop = asyncio.new_event_loop()
+        address = self.loop.run_until_complete(hosts.lease_host())
+        self.ms = MinioServer(self.temp_dir, '127.0.0.1', LogPrefixAdapter(logging.getLogger('minio'), {'prefix': 'minio'}))
+        self.mock_s3 = MockS3Server(address, 2012,
+                                    LogPrefixAdapter(logging.getLogger('s3_mock'), {'prefix': 's3_mock'}))
+
+    def prepare(self) -> None:
+        """
+        Start the S3 mock server and MinIO for the tests.
+        Create a file with environment variables for connecting to them.
+        """
+        tasks = [self.loop.create_task(self.ms.start()), self.loop.create_task(self.mock_s3.start())]
+        self.loop.run_until_complete(asyncio.gather(*tasks))
+        envs = self.ms.get_envs_settings()
+        envs.update(self.mock_s3.get_envs_settings())
+        with open(self.env_file, 'w') as file:
+            for key, value in envs.items():
+                file.write(f"{key}={value}\n")
+
+    def cleanup(self) -> None:
+        """
+        Stop the S3 mock server and MinIO
```

Comment:
By the way, even at this point I don't understand what's the difference between "the S3 mock server" and MinIO. Isn't MinIO also an S3 server? Why do we have both?

Reply:
Moreover, there is a s3 proxy that should run as well. I guess the main reason to have both is to have the ability to inject/simulate errors in s3.